### DISCUSSION TITLE
A retro tkt

### DISCUSSION THREAD

**Discussion by skrutt** [andreisperid/E-TKT#24] [Show and tell]
Starting a thread for my e-tkt build!

It's been in the making for half a year now, or at least in thought. Between waiting for parts, running out of filament and a general lack of free time 😅

Between the time when I "started" and downloaded the files, Andrei produced the PCB and made a bunch of updates to the design. I had already printed all the parts when I noticed, and also have the old esp module. So I'll "press" on, and see where i land.

I have currently assembled most of the PCB, still missing a power section and some cabling. Also not sure what power supply to use when it's ready.. 

Originally my plan was a pretty bare bones implementation, a workhorse integrated into my home assistant instance, even without the screen, since you are probably not looking at the device when you are typing your label.

Then i realized that I actually have several fitting screens on hand! And also that this cute project might be more involved than my first guess, it's always nice to have good feedback when testing stuff.

Today i got the spiffs partition flashed and started on modifying the code to support the smaller screen again, and splitting the display functions into separate files in my fork with a flag for what screen you have. Not quite sure why anyone else would walk my current path by choice, but most of the great open source projects such as marlin have ludicrous support for different hardware so, I'll give it a try.

I'm also planning to try out a microswich for the homing, if it works for homing a 3dprinter it should work here right?

Here are some current pictures:
![PXL_20230331_181638061](https://user-images.githubusercontent.com/3649368/229222891-0ac15c45-db43-4ae7-9a8e-ecf5179146bb.jpg)

![PXL_20230316_204528968](https://user-images.githubusercontent.com/3649368/229223298-bc0d9c26-5c93-427e-bc5f-e3bc4db18905.jpg)

**Comment by andreisperid** [comment]
Hey @skrutt !

Really cool to see another build, I'm really happy. Maybe @sabeechen could post his build in a discussion topic like yours 💡 

PS: I love the color 🍏!

About the power section: since you didn't implement it yet, maybe you can do a few adjustments so you can use a 12v power supply.
- change to a 12v version of the 28BYJ stepper motor (instead of the then-recommended 5v version);
- I didn't test the 7806 linear regulator with 12v input to see if it you heat too much, but if so, perhaps with a heat sink it might do the trick;

### PS: that part of the diagram below is wrong, it should be at max 5v (7v as a stretch), but not 7-12v. But... if you use the 12v stepper version, then 12v is the exact right voltage.

![image](https://user-images.githubusercontent.com/15098003/229513276-2f4b9992-7a4f-47c6-a241-555cc7dadb92.png)


The little screen is so cute... I only changed to a bigger one in the current version because it is too small to read the QR... Anyways, the basic information will fit the smaller, but if you might want to change to a bigger one (128x64px) you would be able to take advantage of the new firmware versions that @sabeechen is contributing to.


Regarding the microswitch, I'm not sure if that would work better than the hall since 3d printers are linear and the switches are placed at the end-stops, differently from the E-TKT, where there is no physical end (cylinder). 

Maybe you would need to change the daisywheel cube to have some relief to trigger the switch, but it might be a mechanical challenge to adjust it perfectly.

Do you have anything against using the hall sensor?

**Reply by skrutt** [reply]
I don't have anything against the hall sensor per se, not apart from being stubborn! And some half baked reasons about adding complexity, cost and more parts. Otoh a switch will actuate at least once per printed letter so I guess a contactless solution might be longer lasting.

I guess we shall find out 🤓

For the mechanical alignment, i think it'll be ok, it has some give in the lever arm so as long as it clicks.. and rotationally I'll adjust in software :)

For the stepper, i just double checked the ones i got and all are 5v.. i guess the only component needing more than 6v are the NEMA stepper. Could it be worth testing that at 6v also? Is the stepper performance/torque critical?

Regarding 7806, i think it will run pretty hot, a TO220 package can dissipate around 1w before overheating, not sure how much the servo needs but say 0.5A would already be 3w. Anything more than that and the heatsink size will skyrocket. But i might have some suitable switchers in stock :)

As an update:
I have mostly done all the conversation to the smaller screen now, I'll try to get some power going tomorrow and get first movement!
I actually had good luck with the QR every time, even with the 128x64 code on the smaller screen when the code was rendered double wide and cropped it read without issue. But apparently the screen has a flicker/scan frequency that sometimes interfere with the shutter speed of the camera. I imagine this could also be an issue on the bigger screen? 
Or it's because of the Noname brand screen i have and other makes might not see that at all.

**Reply by skrutt** [reply]
Nope, A4988 stepper driver is 8v minimum, so not possible with 6v supply only..

**Reply by andreisperid** [reply]
I think my power supply rated "7V" was more than 8V instead (didn't measure under load though), so that makes sense for the A4988 minimum...

But I did use the 7806 with this 8V input and it does not heat too much. But, a 12V input would surely...

About the OLED display, it does "flicker" because of the refresh rate. I remember old CRTs, another retro feature? 😝

**Comment by skrutt** [comment]
![PXL_20230406_150207918](https://user-images.githubusercontent.com/3649368/230482127-d9e0227b-8c9b-45f2-ac9c-51c0187d2d5c.jpg)
An additional update; got a first power supply going! I only had an hour so it was pretty rushed but I managed to get a 7805 connected and wired up, and now all motors are connected and tested! 

There are quite a lot of wiring in this machine, and so far I have spent 0 time to tidy the cabling up so it really looks messy. Just to top it of, the screen fell of after i had verified everything was moving. It'll need some strain relief for sure, the sacrificed jumper duponts have barely any copper in them.

Also saw some erratic behavior, i had not added any capacitance anywhere so hopefully that will be an easy fix.

I was a bit concerned that i had gotten the wrong servo delivered, since i could rotate it several revolutions by hand, but once connected correctly it seems to work without issue.

With regards to the power dissipation of the linear regulator, it seems like it might be fine! I only tested at a conservative 9v but it stayed cool to the touch, so there seems to be quite some headroom to the 100+degrees where the thermal protection kicks in.

I also added a diode between the 5v and 5v-in on the esp board, so USB while powered should be ok. 

And wired up a microswich for homing, but unfortunately I'll need to find some more suitable time for soldering/cleaning up before I get a chance to try coding that part..

**Comment by skrutt** [comment]
Wow, what a ride.. did a first hacky mount of the microswitch yesterday as a proof of concept. i printed a small wedge shape that got superglued into place, and then sacrificed the nema mount because of proximity of the part and suitable hole pattern.. i can say it works! 

The day before though i actually did some first tries with the movement, tried to see if the press and feed mechanics worked but i just could not get any consistent results at all, so i actually caved and placed and order for the hall sensor and magnet, since i just had such a hard time. But then since i still had to wait, i figured that it was really the homing that was missing to go from chaos into order, so i went after the homebrew you see here. 

![PXL_20230408_131611903](https://user-images.githubusercontent.com/3649368/230781571-8eb18a68-e2e8-4051-984d-b21c07f39da2.jpg)

After updating the code to use the digital input, i first had really no results at all. it would always move a bit and then stop in an arbitrary location. 

It was not until i really dug down and did: while(stepchar.move()) that i realized that it must be the input that was misbehaving, and sure enough a quick google did reveal that pin 34 does not have any pullups. DOH. 

After a quick solderjob to add one it was more sensible, and i realized that the char stepper was rotating the wrong way. this would actually go on to be a longer issue as i added a define for inversion of step direction but only for stepsperchar, to some moves where still backwards. 

Long story short, that is now rectified and all moves are clockwise, homing works and i've done quite some testing with calibration!

After finding that the ui for saving the calibration has a small bug, it does not send a valid rest command -> finally i had an excuse to use postman for the first time and i actually managed the first real print! 😄 

![PXL_20230409_151851299](https://user-images.githubusercontent.com/3649368/230782234-5c7068b5-5dca-49f3-916a-31b2c47165de.jpg)

And i apparently placed it upside down in all the excitement! 🙃

**Reply by andreisperid** [reply]
Happy to see some embossed labels! The build is getting nice!

I got wondering how the sound of the embossing would be on every homing, must be cool (I love the microswitch clicks!).  

The stepper direction is something that gave @sabeechen some headaches too. Glad that you figured it out! For the stepper cables, you might just cut them out because the full sized length will surely not fit inside 🤓 

PS: loved the ö for 😲

**Reply by skrutt** [reply]
Well, the ö was meant to be a star 😄
I thought I would try it, turned out ok! I have a swedish character wheel, so that is a future topic.

Do you have any thoughts on this? I have not spotted any provisions for different wheels in the software, though i know you made the wheels changeable. I didn't count yet but guess my wheel has fewer glyphs aswell since there is åäö and also the Danish æ. Maybe it's more Scandinavian than Swedish..

Anyhow, i can read JavaScript but otherwise not very good at "web stuff", so not quite sure how to tackle that. But åäö is printable on keyboard so should also be compatible with no specific handling, except in the character map.

Yeah I really need to sort the cables, some of the jumpers are really low quality and have a tendency to slip out! 

It's very clicky indeed, and seems reliable enough. I have some alignment issues but they seem consistent, and i think might be related to the mechanics.. maybe the shape of the press related to print quality or maybe because I couldn't get the actuator to seat properly, the character pressed is maybe half a character to the left of window center.

In actual printing it's currently a bit silly, i implemented a double tap with a fast search followed by a slow accurate touch on the switch, but only rotating one way that means a whole rotation. And then since the placement is not the same as the original, it first goes to 'j', another half turn, followed by the actual character, could be almost a full turn! So that's like three rotations, for each letter!

I'm for sure gonna look into optimizing that some time, it seems rock solid not loosing steps so i think one homing should do, at least as long as the stepper can stay energized.

**Reply by andreisperid** [reply]
Ah, happy accidents!

So you have the wheel that does not have the symbols but instead these special characters... I used that one in my first prototype too, but when I bought new wheels, I've got the other version with the symbols.

There is a physical possibility for changing the wheel, but the software to support is not developed yet. Would you please open an issue for us to discuss? Maybe on the settings there could be a text form to write all the 43 characters present on the wheel, so they would be supported (today there is a Regex to validate the chars) and transmitted to the machine. As @sabeechen made improvements to support Unicode characters, I guess there would be no big problems on transmitting the åäö, etc.

For the alignment, its the most critical part of the machine, and this is why I've created a fine calibration mode on the newer revision, for both daisywheel alignment and the press angle...

I was wanting to experiment with a single homing too, I guess that could make the printing process at least 10% faster. Anyways, the movement should be always clockwise to avoid entangling the tape on the wheel's teeth. So, for instance, if you go from letter B to A, you still need that almost 360 degree revolution.

If you test this single homing, please let me know how it went!

**Comment by skrutt** [comment]
Will try it! I think it can be much faster than 10%, and there is
additional time to save if the logic for feeding could move straight to the
next position instead of one character at the time.
I have some time tomorrow to try it out again :)

**Comment by skrutt** [comment]
OK, so i did not look into the homing, but i did get the integration into Home assistant working!

Playing with the API did find me some strange behaviour, for one i managed to lock the device in "feeding", where it did not want to return to idle. I did see a note about that in the code so i guess it's not a new thing 😄 

One obvious thing here. since i just put the "command" api value on the dashboard to show what it's doing, there is no idle state! Or rather, i guess there is no "State" at all, which might be nice if building integrations. Not that hard to add in the integration step instead but i do prefer writing c/c++ over YAML! 😱 

Picture from UI:
![bild](https://user-images.githubusercontent.com/3649368/232257968-68322766-d9a0-4ceb-a0a4-e7ecd3bb403b.png)


Here is the configuration details for any interested souls:


```
#Collected details about etkt integration
sensor:
  - platform: rest
    resource: http://192.168.1.154/api/status
    name: E-TKT
    unique_id: ETKT_0001
    json_attributes:
      - busy
      - command
      - align
      - force
    value_template: "{{ value_json.progress }}"
    unit_of_measurement: "%"

input_text:
  etkt_print:
    name: Label text
    initial: Tilde

rest_command:
  etkt_print:
    method: POST
    url: 'http://192.168.1.154/api/task'
    payload: '{"parameter":"tag","value":"  {{ states(''input_text.etkt_print'') }}  "}'
    content_type: 'application/json'
```

And a quick draft of a lovelace card for UI:
```
type: vertical-stack
cards:
  - type: entities
    entities:
      - entity: input_text.etkt_print
      - type: call-service
        name: ' '
        icon: mdi:voice
        action_name: Print It!
        service: rest_command.etkt_print
    title: Skriv etikett
  - type: horizontal-stack
    cards:
      - type: entity
        entity: sensor.e_tkt
        attribute: busy
        name: Upptagen
      - type: entity
        entity: sensor.e_tkt
        attribute: command
        name: Status
      - type: entity
        entity: sensor.e_tkt
        icon: mdi:percent-box
        name: Procent

```

**Comment by andreisperid** [comment]
That's nice!

A few questions:
1. What home assistant are you developing to? Do you have more info on that? (is that one? https://www.home-assistant.io/)
2. What E-TKT firmware are you building over? Because this stuck in "feed", etc, was a problem that happened before @sabeechen implemented the [JSON API](https://github.com/andreisperid/E-TKT/pull/16)

**Comment by skrutt** [comment]
Yes, that's the home assistant. I have a *Home assistant blue" more
specifically, but regardless of hardware should work the same.

I'll have to double check that i have the latest version, otherwise i'll
update. But it's all json for the api

**Comment by sabeechen** [comment]
@andreisperid In a serendipitous twist, Home Assistant happens to be the same automation platform I write that addon for.

I like the idea of being able to print a label from Home Assistant, I was going to make something very much like this once, but I decided I just love that little website too much to give me a reason not to use it :)  
@skrutt I think you could get a meaningful value for "status" if you do a little more to parse out the command, eg
```yaml
sensor:
  - platform: rest
    resource: http://192.168.1.154/api/status
    name: "E-TKT Status"
    unique_id: 17cfa33b-4ff1-4008-96a6-ee60177fd955
    value_template: "{{ value_json.command if value_json.busy else 'idle' }}"
```

Though I think you'd need to the sensor to query/update *constantly* in order for it to ever show anything useful.  I suppose you could write an automation that calls the ```homeassistant.update_entity``` service until it returns to idle any time you print from Home Assistant.  If you're able to reproduce the feeding bug you saw I'd take a look into fixing it.

**Reply by skrutt** [reply]
Thanks! thats a very clean way to template the idle state. 

With regards to the sensor being useful, i feel thats a bit of a.. almost philosophical question. Generally you are either using the device or not using it, and you need to fetch you label either way, so i feel that it's closer to a novelty than real world usefulness. 

BUT i still want it to be working and perfect ofcourse, otherwise i would not be here 😄 

i have not actually looked yet but i would guess there is a way to control the polling rate of the rest sensor? And if you can do that in real time, you should be able to template that into a higher update rate when the print service is called + a reset when idle is detected again. 

Either way, for myself i feel that if i want the extra polish to the integration i might just do a custom component instead, as that would tie both the state and the service into one and provide a (better) place for things like translations and polling rate. Or, perhaps mqtt discovery? I do like the discovery feature... hum. 

With regards to the feeding,  i had 8 commits that i merged yesterday. Quite suprised that it was auto-merged, since i basically changed and split pretty much every file in the src dir. i dont see any changes to the relevant code, but i believe there is a race condition... oh wait, typing here and reading the code simultaneously! When feeding, there is a delay(1000); at the end of displayQRCode();, **after** busy is reset!

```
			busy = false;
			webProgress = 100;
			analogWrite(ledFinish, 0);
			lightChar(0.0f);
			displayQRCode(); // a full second here where a new task can be created
			vTaskDelete(processorTaskHandle); // delete task
```

My guess is that the new request landed just after busy = false;
* a new processor task is created
* the new task sets busy again
* the old task now calls vtaskdelete and deletes the _new_ task, since processorTaskHandle is overwritten
* i guess the old task just sticks arund after that? there is no default case, and i guess "parameter" was either cleared by the new task before deletion, or possibly by the old task the second time around
  * Probably parameter was cleared by the new task, otherwise busy would have been cleared either way

Maybe
```
			webProgress = 100;
			analogWrite(ledFinish, 0);
			lightChar(0.0f);
			displayQRCode();
			busy = false;
			vTaskDelete(NULL); // delete task
```
is good enough? According to the docs, calling vTaskDelete with null will terminate the calling task, so should be closer to the intended result in this case. My tkt is on the healing bench right now awaiting new crimps that actually fit the contact housings so i cant test it right now.

**Reply by skrutt** [reply]
BTW, did any of @andreisperid or @sabeechen look into my [fork](https://github.com/skrutt/E-TKT/tree/master/src)? i feel a little bit silly since i painted myself into my own special corner, and this makes it more difficult for me to contribute upstream. 

If you do have a look, i would love to hear your thoughts on the layout of split files, new defines etc. it should still support a "normal" etkt, but since i dont have one to test on i cannot promise i didnt make any mistakes 👀 

And more stuff also means more clutter so i want to be respectful of the original artists intent. I realize that me wanting to test my ideas might not be of interest to add to official support.

**Reply by andreisperid** [reply]
I'm loving that interaction around the Home Assistant, which I didn't know until now but will figure it out little by little. May be interesting to any other ideas.

What kind of possibilities do you think it brings to E-TKT @sabeechen @skrutt ? I'm particularly interested on voice commands and automation with other chained tasks.

And @skrutt I hadn't looked at your fork yet, thanks for letting me know!

I see you have several things implemented, can you please list their main ideas so we can try to figure out what could be part of the V1.0.0 onwards? If they make sense, we can figure a way to bring them ;)

Regarding the questions related to the Home assistant on the previous post, I will do a bit of research and read it with calm later, but I guess that @sabeechen might already have the answer.

**Reply by sabeechen** [reply]
The freeze up around setting ```busy = [true|false]``` seems likely.  The server should be using semaphores to lock its state between changes, I suspect there are actually many different racy conditions it could run into if you ask for work at an inconvenient time with the API.  I have a TODO somewhere in the codebase for it, ironically.

The biggest advantage for Home Assistant integration I think would just be having a convenient place to print labels, since many are already using it as a kind of dashboard or homepage.  It might be cool for to be able to say "Ok Google,  print a label that says 'Spices'"  though I'm not sure I trust that robot's voice recognition enough to get that right.  Being able to cancel a print in progress over voice might be cool too.

WRT a custom HA component, I think that would be cool.  IIUC mqtt autodiscovery doesn't let you create services, so I don't think that would work, but you can have custom components autoconfigure themselves through network discovery too (not sure how it works precisely). I think it would be good to have the device advertise itself somehow, like over mdns.

Rest sensors only allow a constant polling frequency, but you're separately allowed to call the ```home_assistant.update_entity``` service as often as you want to refresh manually.  I could imagine an automation that triggers when you print a label from the UI that loops on calling that service until the state goes back to idle working pretty well.  Thats how I'd do it, short of having a custom component that handles it more gracefully.  Just make sure it has some kind of back off or early exit so you don't hammer the device constantly with requests and pollute your sensor database if something goes wrong.

You should also know that over the API, the server doesn't check to make sure the label is long enough to be retrieved from inside the machine, you'll need to add padding spaces manually for shorter labels (under 7 chars I think), but thats also likely to change in the future.

@andreisperid If you ever have trouble getting started with Home Assistant don't be afraid to ask me.  Its notoriously difficult for beginners and I've already trialed and errored my way through every problem you could possibly run in to.

**Comment by skrutt** [comment]
Well, the biggest thing i think is the smaller display. To integrate this
in a more structured way i split the display functions out into their own
files, complete with headers and such.

Originally i had a pretty simple idea for that split, cut, paste and
include, but quickly realized that platformio is a bit rigid in the
compilation of source etc so the path of least resistance became a more
formal layout.

Overall the changes:
* Refactor and defines for different display support.
* Reimplementing 32px screen
* Add support for microswich homing, +define
* Define for reversing char stepper direction
* Removal of double negatives for clockwise char stepper
* Minor tweaks, such as renaming optconfig.cpp to .h

**Reply by andreisperid** [reply]
These thing might have potential for a pull request:
* Define for reversing char stepper direction
* Removal of double negatives for clockwise char stepper
* Minor tweaks, such as renaming optconfig.cpp to .h

My only concern is how compatible these changes are to the current code. I've just merged a PR from Stephen recent main changes. Maybe you could update your branch and see if is there any conflict. Then you can make some PRs. What do you think?


PS: I love the small display (though I think it isn't as readable) and the microswitch is a viable alternative for homing, but supporting both in the main project at the same time as the current solutions would be a bit confusing.

**Comment by skrutt** [comment]
Well, I was afraid you would say that, as i’m not sure on how the split into several files will work in a pull request. Or if I can figure out how to.

But overall I *think* it might be possible, git has a certain quality of magic after all. Only way to know is to try 😊

With regards to home assistant, one of my goals for that is to make it more usable to family and friends. If it’s on the same webpage, there is less things to remember. I noticed that the raw api did not add any spaces to the label so I just added two before and two after.

And regarding busy or not.. it a tricky subject that can involve quite some theory. But I think the current API “philosophy” (me reading between the lines of code), is quite fortuitous. The device is either busy, or not, and if the switches in between are cleaned up a bit more this should not really come up.

There might be a few clock cycles between busy = false and the termination of the thread, but that should also be actually OK since the work is done and there is memory to spare.

**Reply by skrutt** [reply]
Ok, so i had a look, and a think, and i think i need to change my approach. 

I'll try to keep upstream compatible features in own branches so that they can be put in a PR, and me-specific ones in their own branch. Also means i kind of need to start over! 

Happy Walpurgis night!

**Reply by andreisperid** [reply]
Nice @skrutt 

@sabeechen brought the initiative of doing a refactoring in #42, which might be a good opportunity for you to sync in with the main E-TKT branch. I'm also learning a lot, it is my first ever opensource interaction :)

PS: didn't know about Walpurgis night, so it seems you are from Germany?

**Reply by andreisperid** [reply]
PS: maybe we can collaborate on the custom daisy wheel issue. As I did with Stephen in #17, I focused on the user interface.
